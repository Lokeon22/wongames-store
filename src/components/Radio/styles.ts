import styled, { css } from "styled-components"

import { RadioProps } from "."

export const Wrapper = styled.main`
  display: flex;
  align-items: center;
  gap: 0.5rem;
`

export const Radio = styled.input`
  ${({ theme }) => css`
    width: 1.9rem;
    height: 1.9rem;
    border-radius: 1.5rem;
    border: 0.2rem solid ${theme.colors.primary};
    appearance: none;
    position: relative;
    outline: none;

    &:focus {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    &:checked {
      background-color: transparent;

      &::before {
        content: "";
        position: absolute;
        top: 0.3rem;
        left: 0.3rem;
        width: 0.9rem;
        height: 0.9rem;
        background-color: ${theme.colors.primary};
        border-radius: 50%;
      }
    }
  `}
`

export const Label = styled.label<Pick<RadioProps, "labelColors">>`
  ${({ theme, labelColors }) => css`
    color: ${theme.colors[labelColors!]};
    line-height: 1;
  `}
`
