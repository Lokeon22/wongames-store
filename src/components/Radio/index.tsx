import React, { InputHTMLAttributes } from "react"
import * as S from "./styles"

type RadioValue = string | ReadonlyArray<string> | number

export type RadioProps = {
  onCheck?: (value?: RadioValue) => void
  label?: string
  labelFor?: string
  labelColors?: "white" | "black"
  value?: RadioValue
} & InputHTMLAttributes<HTMLInputElement>

const Radio = ({
  label,
  onCheck,
  labelFor,
  labelColors = "white",
  value,
  ...props
}: RadioProps) => {
  function onChange() {
    !!onCheck && onCheck(value)
  }

  return (
    <S.Wrapper>
      <S.Radio
        id={labelFor}
        type="radio"
        value={value}
        onChange={onChange}
        {...props}
      />
      {label && (
        <S.Label labelColors={labelColors} htmlFor={labelFor}>
          {label}
        </S.Label>
      )}
    </S.Wrapper>
  )
}

export default Radio
