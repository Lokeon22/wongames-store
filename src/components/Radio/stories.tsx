import { StoryFn, Meta } from "@storybook/react"
import Radio, { RadioProps } from "."

export default {
  title: "Form/Radio",
  component: Radio
} as Meta

export const Default: StoryFn<RadioProps> = (args) => (
  <>
    <div style={{ padding: 10 }}>
      <Radio
        label="High to low"
        labelFor="hightolow"
        value="high"
        name="name"
        defaultChecked
        {...args}
      />
    </div>
    <div style={{ padding: 10 }}>
      <Radio
        label="Low to high"
        labelFor="lowtohigh"
        value="low"
        name="name"
        {...args}
      />
    </div>
  </>
)
