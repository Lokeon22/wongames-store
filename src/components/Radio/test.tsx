import { screen, waitFor } from "@testing-library/react"
import { renderWithTheme } from "../../utils/tests/helpers"

import Radio from "."
import userEvent from "@testing-library/user-event"

describe("<Radio />", () => {
  it("should render with black label", () => {
    const { container } = renderWithTheme(
      <Radio label="radio" labelFor="radio" labelColors="black" />
    )

    expect(screen.getByRole("radio")).toBeInTheDocument()
    expect(screen.getByText(/radio/i)).toHaveStyle({
      color: "#030517"
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it("should render with label white", () => {
    renderWithTheme(<Radio label="radio" labelColors="white" />)

    expect(screen.getByText("radio")).toHaveStyle({
      color: "#FAFAFA"
    })
  })

  it("should be render without label", () => {
    renderWithTheme(<Radio />)

    expect(screen.queryByLabelText(/radio/i)).not.toBeInTheDocument()
  })

  it("should dispatch onCheck when label status changes", async () => {
    const onCheck = jest.fn()
    renderWithTheme(
      <Radio
        label="radio"
        labelFor="radio"
        onCheck={onCheck}
        value="anyvalue"
      />
    )

    expect(onCheck).not.toHaveBeenCalled()
    userEvent.click(screen.getByLabelText("radio"))
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })
    expect(onCheck).toHaveBeenCalledWith("anyvalue")
  })
})
