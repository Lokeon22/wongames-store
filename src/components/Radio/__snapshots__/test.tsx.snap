// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render with black label 1`] = `
.c0 {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.c1 {
  width: 1.9rem;
  height: 1.9rem;
  border-radius: 1.5rem;
  border: 0.2rem solid #F231A5;
  appearance: none;
  position: relative;
  outline: none;
}

.c1:focus {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c1:checked {
  background-color: transparent;
}

.c1:checked::before {
  content: "";
  position: absolute;
  top: 0.3rem;
  left: 0.3rem;
  width: 0.9rem;
  height: 0.9rem;
  background-color: #F231A5;
  border-radius: 50%;
}

.c2 {
  color: #030517;
  line-height: 1;
}

<main
  class="c0"
>
  <input
    class="c1"
    id="radio"
    type="radio"
  />
  <label
    class="c2"
    for="radio"
    labelcolors="black"
  >
    radio
  </label>
</main>
`;
